jenkins/pipelines/ci/tidb-test/tidb_test_ghpr_integration_test.groovy

```diff
@@ -393,6 +420,84 @@ run_with_toolkit_pod {
+                                    curl \${tikv_url} | tar xz bin
+        
+                                    while ! curl --output /dev/null --silent --head --fail \${pd_url}; do sleep 1; done
+                                    curl \${pd_url} | tar xz bin
+        
+                                    mkdir -p ./tidb-src
+                                    curl \${tidb_url} | tar xz -C ./tidb-src
+                                    ln -s \$(pwd)/tidb-src "${ws}/go/src/github.com/pingcap/tidb"
+                                    mv tidb-src/bin/tidb-server ./bin/tidb-server
+                                    """
+                                }
+                            }
+                            try {
+                                timeout(20) {
+                                    sh """
+                                    ps aux
+                                    set +e
+                                    killall -9 -r tidb-server
+                                    killall -9 -r tikv-server
+                                    killall -9 -r pd-server
+                                    rm -rf /tmp/tidb
+                                    rm -rf ./tikv ./pd
+                                    set -e
+                                    
+                                    bin/pd-server --name=pd --data-dir=pd &>pd_${mytest}.log &
+                                    sleep 10
+                                    echo '[storage]\nreserve-space = "0MB"'> tikv_config.toml
+                                    bin/tikv-server -C tikv_config.toml --pd=127.0.0.1:2379 -s tikv --addr=0.0.0.0:20160 --advertise-addr=127.0.0.1:20160 &>tikv_${mytest}.log &
+                                    sleep 10
+                                    if [ -f test.sh ]; then awk 'NR==2 {print "set -x"} 1' test.sh > tmp && mv tmp test.sh && chmod +x test.sh; fi
+
+                                    export TIDB_SRC_PATH=${ws}/go/src/github.com/pingcap/tidb
+                                    export log_level=debug
+                                    TIDB_SERVER_PATH=`pwd`/bin/tidb-server \
+                                    TIKV_PATH='127.0.0.1:2379' \
+                                    TIDB_TEST_STORE_NAME=tikv \
+                                    ${test_cmd}
+                                    """
+                                }
+                            } catch (err) {
+                                sh"""
+                                cat mysql-test.out || true
+                                """
+                                sh """
+                                cat pd_${mytest}.log
+                                cat tikv_${mytest}.log
+                                cat tidb*.log
+                                """
+                                throw err
+                            } finally {
+                                sh """
+                                set +e
```

Comment:
new block codes should not written in groovy pipeline script, prefer to write in a shell, and call in groovy.

Reply:
please look:
- https://github.com/PingCAP-QE/ci/blob/main/pipelines/pingcap/tidb/latest/ghpr_check2.groovy#L143
- https://github.com/PingCAP-QE/ci/blob/main/scripts/pingcap/tidb/explaintest.sh

Reply:
This part of the code was just copied from another groovy pipeline.
I am not a professional developer of test pipelines. 
If you want clean code, shouldn't the EE team provide a function that can be passed in Docker Image names to adapt to different languages? Instead of simply restricting from writing shells in Groovy.

Reply:
said by icemap: it will solved later in other pull request.